### ISSUE TITLE
BufWriter's inner does not need to be an Option

### ISSUE THREAD

**Issue by kestein** [rust-lang/rust#36319]
In the implementation of [BufWriter](https://doc.rust-lang.org/src/std/up/src/libstd/io/buffered.rs.html#300-307) the option type for the inner writer does not seem to be necessary. There is no way to initialize the inner writer to be None and the value is unwrap()'d anywhere the value of it is needed.

**Comment by TimNN** [comment]
The `Option` is necessary to allow for a `into_inner` implementation in the presence of a `Drop` implementation.

**Comment by apasel422** [comment]
@TimNN `into_inner` could be rewritten to use `ptr::read` and `mem::forget`.

**Comment by TimNN** [comment]
@apasel422: Yeah, you're right, so let me rephrase: An `Option` is necessary for a *safe* `into_inner` implementation. 

And personally I don't think the `unsafe` is worth it here, since whatever cost is associated with the `Option` is probably negligible when compared to the cost of the actual I/O (and, IMO, the less `unsafe` the better).

**Comment by rkruppe** [comment]
A benchmark would be good, but if there is any measurable gain to be had, then this is *exactly* (one of the reasons) why `unsafe` exists. It might be sad that this pattern can't be expressed safely, but that's a grievance with the language, not a mark against the pattern, modulo the practical concerns of how sure one is about the correctness of the implementation. But I'm not super worried about the safety since the `unsafe` code would be rather small, simple, and contained.

**Comment by frewsxcv** [comment]
Anyone have ideas for a benchmark for this?

**Comment by Mark-Simulacrum** [comment]
I'm going to go ahead and close as this is a pretty ugly change based on the couple PRs that tried to fix this and the additional complexity isn't really worth it (no benchmarks are able to show significant wins).
